### PR TITLE
	Specialized Headers impl backed by Akka Http

### PR DESCRIPTION
**Pull request by ktoso**
Two optimisations.

One: to check that if contentLength is given but only if > 0 then we need to materialize things
Two: To avoid copying all headers into strings all the time but only when really we have to.
The special cased ContentLength, ContentType and uri help a bit too.

I got a few failures locally but did not think they're related, so want to check on travis too while I look into those.

### CODE REVIEW

File: framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala [hunk 1]
```diff
@@ -100,20 +101,21 @@ private[server] class AkkaModelConversion(
...
     }
 
     var requestUri: String = null
     request.headers.foreach {
       case `Raw-Request-URI`(u) =>
         requestUri = u
-        false // exclude the synthetic header from effective headers
-      case _ => true
+      case e: `Transfer-Encoding` =>
+        knownTransferEncoding = Some(e.value())
```
Comment (ktoso) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: we find it eagerly since we do a foreach anyway - we could also temrinate the foreach eagerly once we found those two

**Comment by ktoso** [comment]
This patch should yield another 5% or so perf improvement, fine tuning it still though

File: framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala (hunk 1 above)
Comment (jrudolph) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: Even with filter it was way below 1%, so no further optimization necessary on that one.

Comment (jrudolph) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: There's will never be a `Transfer-Encoding` header in akka-http so no need to match here, the previous solution using `isChunked` was fine. For `get` just return a `Some("chunked")` if `isChunked`.

File: framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala [hunk 2]
```diff
@@ -258,3 +260,67 @@ private[server] class AkkaModelConversion(
...
+final case class AkkaHeadersWrapper(
+    request: HttpRequest,
+    knownContentLength: Option[String],
+    hs: immutable.Seq[HttpHeader],
+    knownTransferEncoding: Option[String],
+    uri: String
+) extends Headers(null) {
+
+  override def headers: Seq[(String, String)] =
+    if (_headers ne null) _headers
```
Comment (jrudolph) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: Why is that needed?

File: framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala [hunk 3]
```diff
@@ -258,3 +260,67 @@ private[server] class AkkaModelConversion(
...
+
+final case class AkkaHeadersWrapper(
+    request: HttpRequest,
+    knownContentLength: Option[String],
+    hs: immutable.Seq[HttpHeader],
+    knownTransferEncoding: Option[String],
+    uri: String
+) extends Headers(null) {
+
+  override def headers: Seq[(String, String)] =
```
Comment (jrudolph) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: Should probably be a `lazy val`.

Comment (ktoso) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: it's basically a lazy val but without the locks, stores itself in the field that's otherwise empty

File: framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala (hunk 2 above)
Comment (ktoso) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: it's like a lazy val basically

File: framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala (hunk 1 above)
Comment (ktoso) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: Hm, I see. I was not sure what about the other values of transfer encoding, but we dont really care - all we care about is that its chunked hm

**Comment by ktoso** [comment]
I'll revert the transfer encoding change

**Comment by ktoso** [comment]
Cleaned up the PR

File: framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala [hunk 4]
```diff
@@ -258,3 +259,67 @@ private[server] class AkkaModelConversion(
...
+    knownContentLength: Option[String],
+    hs: immutable.Seq[HttpHeader],
+    isChunked: Option[String],
+    uri: String
+) extends Headers(null) {
+
+  override def headers: Seq[(String, String)] =
+    if (_headers ne null) _headers
+    else {
+      _headers = hs.map(h => h.name() -> h.value) // TODO should we mark _headers volatile? It's harmless to re-gen them again
```
Comment (jrudolph) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: As I say :) Use a `lazy val` and there won't be any questions. The lock will be much cheaper than rendering the headers, so probably won't matter in practice.

Comment (ktoso) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: Ok, might as well :)

File: framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala [hunk 5]
```diff
@@ -258,3 +259,67 @@ private[server] class AkkaModelConversion(
...
+    copy(hs = this.hs ++ raw(headers))
+
+  override def apply(key: String): String =
+    get(key).getOrElse(throw new RuntimeException(s"Header with name ${key} not found!"))
+
+  override def get(key: String): Option[String] =
+    key match {
+      case HeaderNames.CONTENT_LENGTH => knownContentLength
+      case HeaderNames.TRANSFER_ENCODING => isChunked
+      case HeaderNames.CONTENT_TYPE => Option(request.entity.contentType.value)
```
Comment (jrudolph) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: Why wrapped in `Option`? It will never be `null`. What you can do instead is checking before if the entity `isKnownEmpty` and return `None` in that case.

File: framework/src/play/src/main/scala/play/api/mvc/Headers.scala [hunk 6]
```diff
@@ -6,6 +6,7 @@ package play.api.mvc
 import java.util.Locale
 
 import play.core.utils.CaseInsensitiveOrdered
+import play.mvc.Http.HeaderNames
```
Comment (jrudolph) [inline framework/src/play/src/main/scala/play/api/mvc/Headers.scala]: No changes needed here for now.

File: framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala (hunk 5 above)
Comment (ktoso) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: Option since that's the API I need to conform with here (can just make is `Some()` sure)

Comment (ktoso) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: > What you can do instead is checking before if the entity isKnownEmpty and return None in that case.

Good one

File: framework/src/play/src/main/scala/play/api/mvc/RequestHeader.scala [hunk 7]
```diff
@@ -160,8 +166,9 @@ trait RequestHeader {
    */
   def hasBody: Boolean = {
     import HeaderNames._
-    headers.get(CONTENT_LENGTH).isDefined || headers.get(TRANSFER_ENCODING).isDefined
+    headers.get(CONTENT_LENGTH).exists(_gtZero) || headers.get(TRANSFER_ENCODING).isDefined
   }
+  private val _gtZero: String => Boolean = _.toLong > 0
```
Comment (jrudolph) [inline framework/src/play/src/main/scala/play/api/mvc/RequestHeader.scala]: If we do the `isKnownEmpty` check above, no changes are needed here. Otherwise, you also need to check that `toLong` doesn't fail.

File: framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala (hunk 5 above)
Comment (ktoso) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: Wait... dropping content type if no entity present does not sound that safe actually hm...

File: framework/src/play/src/main/scala/play/api/mvc/RequestHeader.scala (hunk 7 above)
Comment (ktoso) [inline framework/src/play/src/main/scala/play/api/mvc/RequestHeader.scala]: I'm not sure we need any aditional code for the isKnownEmpty... won't these cover it "automagically"? Since it remains None until we have some 

```
    var knownContentLength: Option[String] = None
    request.entity match {
      case HttpEntity.Strict(_, data) =>
        if (request.method.requestEntityAcceptance == RequestEntityAcceptance.Expected || data.nonEmpty) 
          knownContentLength = Some(data.length.toString)
      case HttpEntity.Default(_, cLength, _) =>
        if (request.method.requestEntityAcceptance == RequestEntityAcceptance.Expected || cLength > 0) 
          knownContentLength = Some(cLength.toString)
```

So if it's not known it'll return `None`, so yeah then the chunked check kicks in.

If we were to access specific "are you known empty" from here than that'd be optimal sure, but that's a larger change to `Headers`. Unless I'm misunderstanding your intent

Comment (jrudolph) [inline framework/src/play/src/main/scala/play/api/mvc/RequestHeader.scala]: I might have confused things here. Maybe `_gtZero` just isn't currently needed because of the block you referenced.

Comment (ktoso) [inline framework/src/play/src/main/scala/play/api/mvc/RequestHeader.scala]: Right, it's not actually needed now. Should not add impact though since it'll be a none there, shall I keep?

**Comment by ktoso** [comment]
Please review when you have the time @richdougherty

File: framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala [hunk 8]
```diff
@@ -258,3 +259,65 @@ private[server] class AkkaModelConversion(
...
+
+final case class AkkaHeadersWrapper(
+    request: HttpRequest,
+    knownContentLength: Option[String],
+    hs: immutable.Seq[HttpHeader],
+    isChunked: Option[String],
+    uri: String
+) extends Headers(null) {
+
+  override lazy val headers: Seq[(String, String)] =
```
Comment (ktoso) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: I would actually prefer the if == null style rather than the lazy val but let's see until it shows up in benchmarking

**Comment by ktoso** [comment]
Made it compile in all modules finally

File: framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala [hunk 9]
```diff
@@ -69,8 +73,8 @@ private[server] class AkkaModelConversion(
         headers),
       request.method.name,
       new RequestTarget {
-        override lazy val uri: URI = new URI(uriString)
-        override lazy val uriString: String = _uriString
+        override lazy val uri: URI = new URI(headers.uri)
+        override lazy val uriString: String = headers.uri
```
Comment (richdougherty) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: If `headers.uri` is cheap then you could make this method a `def`.

File: framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala [hunk 10]
```diff
@@ -83,37 +87,34 @@ private[server] class AkkaModelConversion(
...
-        requestUri
+    var requestUri: String = null
+    request.headers.foreach {
+      case `Raw-Request-URI`(u) =>
+        requestUri = u
+      case e: `Transfer-Encoding` =>
+        isChunked = Some(e.value())
+      case _ => // continue
+    }
+    if (requestUri eq null) requestUri = request.uri.toString() // fallback value
```
Comment (richdougherty) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: Sorry if this is a dumb question, I'm not sure what a `Raw-Request-URI` is. Should this value feed into the `RequestTarget`'s URI value? At the moment the `RequestTarget` is getting the URI value from `request.uri`.

File: framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala [hunk 11]
```diff
@@ -258,3 +259,63 @@ private[server] class AkkaModelConversion(
...
+    get(key).getOrElse(throw new RuntimeException(s"Header with name ${key} not found!"))
+
+  override def get(key: String): Option[String] =
+    key match {
+      case HeaderNames.CONTENT_LENGTH => knownContentLength
+      case HeaderNames.TRANSFER_ENCODING => isChunked
+      case HeaderNames.CONTENT_TYPE => Some(request.entity.contentType.value)
+      case _ =>
+        val lower = key.toLowerCase(Locale.ROOT)
+        hs.collectFirst({ case h if h.lowercaseName == lower => h.value })
```
Comment (richdougherty) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: Hopefully iterating over a `Seq` isn't too slow in practice. In most workloads I'd guess that a `Seq` is probably faster than a `Map` because it's cheaper to construct.

File: framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala [hunk 12]
```diff
@@ -258,3 +259,63 @@ private[server] class AkkaModelConversion(
...
+  // note that these are rarely used, mostly just in tests
+  override def replace(headers: (String, String)*): Headers = {
+    val replaced = hs.filterNot(h => headers.exists(rm => h.is(rm._1))) ++ raw(headers)
+    copy(hs = replaced)
+  }
+
+  override def equals(other: Any): Boolean = {
+    other match {
+      case that: AkkaHeadersWrapper => that.request == this.request
+      case _ => false
```
Comment (richdougherty) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: Do we ever need to compare to a non-Akka `Headers` object? I guess not?

File: framework/src/play/src/main/scala/play/api/mvc/RequestHeader.scala (hunk 7 above)
Comment (richdougherty) [inline framework/src/play/src/main/scala/play/api/mvc/RequestHeader.scala]: If you want to keep it then I think move it into the `object RequestHeader` so we don't create one for each `RequestHeader`.

File: framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala (hunk 9 above)
Comment (ktoso) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: good idea

File: framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala (hunk 12 above)
Comment (ktoso) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: Right, I think that won't happen in practice since you're either netty or the akka backend

File: framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala (hunk 10 above)
Comment (ktoso) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: There are no dumb questions! 😉 
`Raw-Request-URI` is used to give access to the "not url decoded" url. It seems as if Play needed that? Though it seems it is currently disabled actually (the setting is not set): https://github.com/playframework/playframework/blob/master/framework/src/play-akka-http-server/src/main/resources/reference.conf

So perhaps we can avoid doing the Raw filtering completely...

There's some docs here http://doc.akka.io/docs/akka-http/10.0.4/scala/http/common/http-model.html#obtaining-the-raw-request-uri

File: framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala (hunk 11 above)
Comment (ktoso) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: Should be ok

**Comment by ktoso** [comment]
Sorry I did not get to finishing this before my move, I also har surface-work + conference during that week so barely survived that, not mentioning working on these :~

Will get to it next week, unless someone can polish the last remaining thing in it?

File: framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala (hunk 10 above)
Comment (jrudolph) [inline framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala]: The setting is set here: https://github.com/playframework/playframework/blob/1e56a9f560dabfcc5368027a59d61bc90466462b/framework/src/play-akka-http-server/src/main/scala/play/core/server/AkkaHttpServer.scala#L81-L81.

The alternative would be to make the akka-http URI model work for play but maybe let's not go there for now.

**Comment by jrudolph** [comment]
I fixed the three failing test classes, will push an update soon.

**Comment by jrudolph** [comment]
Couldn't push to your branch, @ktoso, so I created https://github.com/playframework/playframework/pull/7222 with the fixed version.

**Comment by jrudolph** [comment]
This one can be closed (but I don't have the rights).

**Comment by ktoso** [comment]
Ok, closing this one then. I'll give you write access in case you'd need it any time else
